Question_Categorization/clustering_and_tagging_classes.py

```diff
@@ -286,13 +287,46 @@ def map_tags(self, tag_dict):
         '''
         try:
             mapped_dict = {}
+            lemmatizer = WordNetLemmatizer()
+            
             for tag_tuple in tag_dict:
-                tag_string = tag_tuple[0] #Replace with a mapping algorithm
+
+                #Create an embedding for the set of questions belonging to the tuple of tags.
+                question_string = ' '.join([question for question in tag_dict[tag_tuple]])
+                question_embedding = embed([question_string])
+
+                #Create a list of tags, where each element is a list contained the lemmatized forms of words present in the tag.
+                tag_list = [[lemmatizer.lemmatize(word) for word in tag.split(' ')] for tag in tag_tuple]
+
+                #Create embeddings for the lemmatized tags.
+                tag_embedding_list = [' '.join(tag) for tag in tag_list]
+                tag_embeddings = embed(tag_embedding_list)
+                
+                #Compute the scores for each word present in the list of tags, based on their frequency in English.
+                word_list = list(set([word for tag in tag_list for word in tag]))
+                word_scores = {}
+                for word in word_list: 
+                    word_scores[word] = (word_frequency(word, 'en') * 100) + 1
+
+                #Compute scores for each tag in the list.
+                tag_scores = []
+                for tag in tag_list:
+                    #Compute the similarity between the tag and the question set.
+                    tag_similarity = 1 - cosine(question_embedding, tag_embeddings[tag_list.index(tag)])
+                    #Compute the average frequency in English of the words present in the tag.
+                    denom = np.array([word_scores[word] for word in tag]).mean()
+                    tag_score = tag_similarity/denom
```

Comment:
Intuitively, I like the metric you chose! It makes sense to prioritize both being a good match and being a descriptive tag.